analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/InvalidCastToInterface.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Collections.Generic;
+
+public interface IMyInterface { }
+public interface IMyInterface2 { }
+public interface IMyInterface3 : IMyInterface { }
+public interface IMyInterface4 : IMyInterface { }
+public interface IMyInterface5 { }
+
+public class Implementer : IMyInterface { }
+public class Implementer5 : IMyInterface5 { }
+
+public class MyClass1 { }
+public class MyClass2 { }
+public class MyClass3 : MyClass2, IMyInterface { }
+public class MyClass4 { }
+
+public class InvalidCastToInterface
+{
+    public class Nested : MyClass4, IDisposable
+    {
+        public void Dispose() { }
+    }
+
+    static void Main()
+    {
+        var myclass1 = new MyClass1();
+        var x = (IMyInterface)myclass1; // Noncompliant {{Review this cast; in this project there's no type that extends 'MyClass1' and implements 'IMyInterface'.}},   this part is not based on SE
+//               ^^^^^^^^^^^^
+        var a = (IMyInterface5)x;       // Noncompliant {{Review this cast; in this project there's no type that implements both 'IMyInterface' and 'IMyInterface5'.}}, this part is not based on SE
+        x = myclass1 as IMyInterface;
+        bool b = myclass1 is IMyInterface;
+
+        var arr = new MyClass1[10];
+        var arr2 = (IMyInterface[])arr;
+
+        var myclass2 = new MyClass2();
+        var y = (IMyInterface)myclass2;
+
+        IMyInterface i = new MyClass3();
+        var c = (IMyInterface2)i; // Compliant, because IMyInterface2 doesn't have concrete implementation
+        IMyInterface4 ii = null;
+        var d = (IMyInterface2)i; // Compliant
+        var e = (IMyInterface3)i;
+
+        var o = (object)true;
+        e = (IMyInterface3)o;
+
+        var coll = (IEnumerable<int>)new List<int>();
+        var z = (IDisposable)new MyClass4();
+        var w = (IDisposable)(new Node());
+    }
+}
+
+public class DerivedNode : MiddleNode, IDisposable
+{
+    public void Dispose() { }
+}
+
+public class MiddleNode : Node { }
+public class Node { }
+
+public class MyClass
+{
+    public double? D { get; set; } = 1.001;
+}
+
+public class NullableTest
+{
+    public void Test1()
+    {
+        int? i1 = null;
+        var ii = (int)i1; // FIXME Non-compliant {{Nullable is known to be empty, this cast throws an exception.}}, this part is not based on SE
```

Comment:
Surely this is SE based, right?

Reply:
Yep, I'll apply this in https://github.com/SonarSource/sonar-dotnet/pull/7307